python/ray/tests/test_kill_subprocesses.py

```diff
@@ -0,0 +1,122 @@
+
+
+@pytest.fixture
+def enable_subreaper():
+    os.environ["RAY_kill_child_processes_on_worker_exit_with_raylet_subreaper"] = "true"
+    yield
+    del os.environ["RAY_kill_child_processes_on_worker_exit_with_raylet_subreaper"]
+
+
+def sleep_forever():
+    while True:
+        time.sleep(10000)
+
+
+def get_process_info(pid):
+    # may raise psutil.NoSuchProcess
+    process = psutil.Process(pid)
+    return {
+        "PID": process.pid,
+        "Name": process.name(),
+        "Status": process.status(),
+        "CPU Times": process.cpu_times(),
+        "Memory Info": process.memory_info(),
+    }
+
+
+@ray.remote
+class BedMaker:
+    def make_sleeper(self):
+        p = multiprocessing.Process(target=sleep_forever)
+        p.start()
+        return p.pid
+
+    def spawn_daemon(self):
+        import subprocess
+
+        # Spawns a bash script (shell=True) that starts a daemon process
+        # which sleeps 1000s. The bash exits immediately, leaving the daemon
+        # running in the background. We don't want to kill the daemon when
+        # the actor is alive; we want to kill it after the actor is killed.
+        command = "nohup sleep 1000 >/dev/null 2>&1 & echo $!"
+        output = subprocess.check_output(command, shell=True, text=True)
+        return int(output.strip())
+
+    def my_pid(self):
+        return os.getpid()
+
+
+@pytest.mark.skipif(
+    sys.platform != "linux",
+    reason="Orphan process killing only works on Linux.",
+)
+def test_ray_kill_can_kill_subprocess(shutdown_only):
+    """
+    This works becuase of kill_child_processes_on_worker_exit.
+    Even if kill_child_processes_on_worker_exit_with_raylet_subreaper
+    is not set, the worker will still kill its subprocesses.
+    """
+    ray.init()
+    b = BedMaker.remote()
+    pid = ray.get(b.make_sleeper.remote())
+
+    # ray.kill can kill subprocesses.
+    logger.info(get_process_info(pid))  # shows the process
+    ray.kill(b)
+    time.sleep(1)
+    with pytest.raises(psutil.NoSuchProcess):
+        logger.info(get_process_info(pid))  # subprocess killed
+
+
+@pytest.mark.skipif(
+    sys.platform != "linux",
+    reason="Orphan process killing only works on Linux.",
+)
+def test_sigkilled_worker_can_kill_subprocess(enable_subreaper, shutdown_only):
+    ray.init()
+    # sigkill'd actor can't kill subprocesses
+    b = BedMaker.remote()
+    pid = ray.get(b.make_sleeper.remote())
+    actor_pid = ray.get(b.my_pid.remote())
+
+    logger.info(get_process_info(pid))  # shows the process
+    psutil.Process(actor_pid).kill()  # sigkill
+    time.sleep(11)  # unowned processes are killed every 10s.
+    with pytest.raises(psutil.NoSuchProcess):
+        logger.info(get_process_info(pid))  # subprocess killed
+
+
+@pytest.mark.skipif(
+    sys.platform != "linux",
+    reason="Orphan process killing only works on Linux.",
+)
+def test_daemon_processes_not_killed_until_actor_dead(enable_subreaper, shutdown_only):
```

Comment:
Can you add a test where adding signal can reap grandchildren?

Reply:
no need, reversed to core worker to default-reap all children.

Reply:
got it! please make sure you added a test case for reap. (and document how to diable automatic reap)